src/core/containers.h

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) 2014-2015 Dropbox, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef PYSTON_CORE_CONTAINERS_H
+#define PYSTON_CORE_CONTAINERS_H
+
+
+namespace pyston {
+
+template <class T, class TAllocator = std::allocator<T>> class SmallStackVector {
+private:
+    size_t capacity;
+    size_t _size;
+    T data[8];
+    T* datap;
+    TAllocator alloc;
+
+public:
+    SmallStackVector() {
+        capacity = sizeof(data) / sizeof(data[0]);
+        _size = 0;
+        datap = &data[0];
+    }
+    SmallStackVector(const SmallStackVector& vec) {
+        capacity = sizeof(data) / sizeof(data[0]);
+        _size = vec._size;
+        datap = &data[0];
+        maybe_resize();
+        for (int i = 0; i < _size; i++) {
+            datap[i] = vec.datap[i];
+        }
+    }
+    SmallStackVector(size_t _sz) {
+        capacity = sizeof(data) / sizeof(data[0]);
+        _size = _sz;
+        datap = &data[0];
+        maybe_resize();
+        memset(datap, 0, sizeof(T) * _size);
+    }
+    SmallStackVector(std::initializer_list<T> init) {
+        capacity = 16;
+        _size = init.size();
+        datap = &data[0];
+
+        maybe_resize();
+
+        int i = 0;
+        for (auto e : init) {
+            alloc.construct(&datap[i++], std::move(e));
+        }
+    }
+    ~SmallStackVector() {
+        if (datap != &data[0])
+            alloc.deallocate(datap, 1);
```

Comment:
I think the this should be ```alloc.deallocate(data, capacity)```
Also I think we don't destroy the objects inside...